### PR TITLE
Error handling

### PR DESCRIPTION
**Pull request by faytey**
A draft of the error-handling techniques used for testing purposes, kindly let me know if the method used so far is acceptable. Thanks

### CODE REVIEW

File: src/contracts/yas_pool.cairo [hunk 1]
```diff
@@ -25,7 +25,8 @@ trait IYASPool<TContractState> {
 
 #[starknet::contract]
 mod YASPool {
-    use super::IYASPool;
+    use core::result::ResultTrait;
+use super::IYASPool;
```
Comment (uri-99) [inline src/contracts/yas_pool.cairo:29]: format error. You should run `make fmt` before pushing

File: src/libraries/tick.cairo [hunk 2]
```diff
@@ -74,7 +73,8 @@ trait ITick<TContractState> {
 
 #[starknet::contract]
 mod Tick {
-    use super::{ITick, Info};
+    use core::result::ResultTrait;
+use super::{ITick, Info};
```
Comment (uri-99) [inline src/libraries/tick.cairo:77]: format error. You should run `make fmt` before pushing

File: src/libraries/tick_bitmap.cairo [hunk 3]
```diff
@@ -10,7 +10,8 @@ trait ITickBitmap<TContractState> {
 
 #[starknet::contract]
 mod TickBitmap {
-    use super::ITickBitmap;
+    use core::result::ResultTrait;
+use super::ITickBitmap;
```
Comment (uri-99) [inline src/libraries/tick_bitmap.cairo:14]: format error. You should run `make fmt` before pushing

File: src/tests/test_libraries/test_bit_math.cairo [hunk 4]
```diff
@@ -1,81 +1,131 @@
 mod BitMathTests {
     mod MostSignificantBit {
-        use integer::BoundedInt;
+        use core::result::ResultTrait;
+use integer::BoundedInt;
```
Comment (uri-99) [inline src/tests/test_libraries/test_bit_math.cairo:4]: format error. You should run `make fmt` before pushing

File: src/tests/test_libraries/test_bit_math.cairo [hunk 5]
```diff
@@ -1,81 +1,131 @@
...
         fn msb_number_zero() {
-            let ret = most_significant_bit(0);
+            let ret = most_significant_bit(0).unwrap();
         }
     }
 
     mod LeastSignificantBit {
-        use integer::BoundedInt;
+        use core::result::ResultTrait;
+use integer::BoundedInt;
```
Comment (uri-99) [inline src/tests/test_libraries/test_bit_math.cairo:68]: format error. You should run `make fmt` before pushing

**Review by uri-99 — CHANGES_REQUESTED** [review]
some format errors. You should run `make fmt` before pushing

File: src/tests/test_libraries/test_bit_math.cairo [hunk 6]
```diff
@@ -1,81 +1,131 @@
 mod BitMathTests {
     mod MostSignificantBit {
-        use integer::BoundedInt;
+        use core::result::ResultTrait;
+use integer::BoundedInt;
         use yas::libraries::bit_math::BitMath::most_significant_bit;
 
         #[test]
-        fn msb_happy_path() {
+        fn msb_happy_path() -> Result<u8, felt252> {
```
Comment (uri-99) [inline src/tests/test_libraries/test_bit_math.cairo:8]: `#[test]` should contain `#[available_gas()]` attribute 
```suggestion
        #[available_gas(200000000)]
        fn msb_happy_path() -> Result<u8, felt252> {
```

File: src/tests/test_libraries/test_bit_math.cairo [hunk 7]
```diff
@@ -1,81 +1,131 @@
...
+            // assert(most_significant_bit(1) == 0, 'msb should be 0');
+            if !(most_significant_bit(1).expect('msb errored') == 0 ) {
+                return Result::Err('msb should be 0');
+            }
+            return most_significant_bit(1);
         }
 
         #[test]
-        fn msb_larger_number() {
+        fn msb_larger_number() -> Result<u8, felt252> {
```
Comment (uri-99) [inline src/tests/test_libraries/test_bit_math.cairo:18]: `#[test]` should contain `#[available_gas()]` attribute

File: src/tests/test_libraries/test_bit_math.cairo [hunk 8]
```diff
@@ -1,81 +1,131 @@
...
+            // assert(most_significant_bit(128) == 7, 'msb should be 7');
+            if !(most_significant_bit(128).expect('msb errored') == 7 ) {
+                return Result::Err('msb should be 7');
+            }
+            return most_significant_bit(128);
         }
 
         #[test]
-        fn msb_bigger_number() {
+        fn msb_bigger_number() -> Result<u8, felt252> {
```
Comment (uri-99) [inline src/tests/test_libraries/test_bit_math.cairo:28]: `#[test]` should contain `#[available_gas()]` attribute

File: src/tests/test_libraries/test_bit_math.cairo [hunk 9]
```diff
@@ -1,81 +1,131 @@
...
+                return Result::Err('msb should be 19');
+            }
+            return most_significant_bit(1000000);
+        
         }
 
         #[test]
-        fn msb_maximum_256() {
-            assert(most_significant_bit(BoundedInt::max()) == 255, 'msb should be 255');
+        fn msb_maximum_256() -> Result<u8, felt252> {
```
Comment (uri-99) [inline src/tests/test_libraries/test_bit_math.cairo:39]: `#[test]` should contain `#[available_gas()]` attribute

File: src/tests/test_libraries/test_bit_math.cairo [hunk 10]
```diff
@@ -1,81 +1,131 @@
...
+            if !(most_significant_bit(BoundedInt::max()).expect('msb errored') == 255) {
+                return Result::Err('msb should be 255');
+            }
+            return most_significant_bit(BoundedInt::max());
+        
         }
 
         #[test]
-        fn msb_random_number() {
+        fn msb_random_number() -> Result<u8, felt252> {
```
Comment (uri-99) [inline src/tests/test_libraries/test_bit_math.cairo:49]: `#[test]` should contain `#[available_gas()]` attribute

File: src/tests/test_libraries/test_bit_math.cairo [hunk 11]
```diff
@@ -1,81 +1,131 @@
...
+                return Result::Err('msb should be 13');
+            }
+            return most_significant_bit(12345);
+        
         }
 
         #[test]
-        #[should_panic]
+        // #[should_panic]
         fn msb_number_zero() {
```
Comment (uri-99) [inline src/tests/test_libraries/test_bit_math.cairo:61]: `#[test]` should contain `#[available_gas()]` attribute

File: src/tests/test_libraries/test_bit_math.cairo [hunk 12]
```diff
@@ -1,81 +1,131 @@
...
 
     mod LeastSignificantBit {
-        use integer::BoundedInt;
+        use core::result::ResultTrait;
+use integer::BoundedInt;
         use yas::libraries::bit_math::BitMath::least_significant_bit;
 
         #[test]
-        fn lsb_happy_path() {
+        fn lsb_happy_path() -> Result<u8, felt252> {
```
Comment (uri-99) [inline src/tests/test_libraries/test_bit_math.cairo:72]: `#[test]` should contain `#[available_gas()]` attribute

File: src/tests/test_libraries/test_bit_math.cairo [hunk 13]
```diff
@@ -1,81 +1,131 @@
...
+            if !(least_significant_bit(1).expect('lsb errored') == 0) {
+                return Result::Err('lsb should be 0');
+            }
+            return least_significant_bit(1);
+        
         }
 
         #[test]
-        fn lsb_larger_number() {
+        fn lsb_larger_number() -> Result<u8, felt252> {
```
Comment (uri-99) [inline src/tests/test_libraries/test_bit_math.cairo:83]: `#[test]` should contain `#[available_gas()]` attribute

File: src/tests/test_libraries/test_bit_math.cairo [hunk 14]
```diff
@@ -1,81 +1,131 @@
...
+            if !(least_significant_bit(128).expect('lsb errored') == 7) {
+                return Result::Err('lsb should be 7');
+            }
+            return least_significant_bit(128);
+        
         }
 
         #[test]
-        fn lsb_bigger_number() {
+        fn lsb_bigger_number() -> Result<u8, felt252> {
```
Comment (uri-99) [inline src/tests/test_libraries/test_bit_math.cairo:94]: `#[test]` should contain `#[available_gas()]` attribute

File: src/tests/test_libraries/test_bit_math.cairo [hunk 15]
```diff
@@ -1,81 +1,131 @@
...
+            if !(least_significant_bit(1000000).expect('lsb errored') == 6) {
+                return Result::Err('lsb should be 6');
+            }
+            return least_significant_bit(1000000);
+        
         }
 
         #[test]
-        fn lsb_maximum_256() {
+        fn lsb_maximum_256() -> Result<u8, felt252> {
```
Comment (uri-99) [inline src/tests/test_libraries/test_bit_math.cairo:105]: `#[test]` should contain `#[available_gas()]` attribute

File: src/tests/test_libraries/test_bit_math.cairo [hunk 16]
```diff
@@ -1,81 +1,131 @@
...
+            if !(least_significant_bit(BoundedInt::max()).expect('lsb errored') == 0) {
+                return Result::Err('lsb should be 0');
+            }
+            return least_significant_bit(BoundedInt::max());
+        
         }
 
         #[test]
-        fn lsb_random_number() {
+        fn lsb_random_number() -> Result<u8, felt252> {
```
Comment (uri-99) [inline src/tests/test_libraries/test_bit_math.cairo:116]: `#[test]` should contain `#[available_gas()]` attribute

File: src/tests/test_libraries/test_bit_math.cairo [hunk 17]
```diff
@@ -1,81 +1,131 @@
...
+                return Result::Err('lsb should be 0');
+            }
+            return ret;
+        
         }
 
         #[test]
-        #[should_panic]
+        // #[should_panic]
         fn lsb_number_zero() {
```
Comment (uri-99) [inline src/tests/test_libraries/test_bit_math.cairo:129]: `#[test]` should contain `#[available_gas()]` attribute

**Review by uri-99 — CHANGES_REQUESTED** [review]
some missing `#[available_gas()]` attributes in `#[tests]`. It is important to evaluate gas consumption, to ensure there are no deadlocks.

**Review by uri-99 — CHANGES_REQUESTED** [review]
Tests should `assert()` different values regarding the functionality being tested; to ensure the functionality fails when it should fail and also to ensure that it doesn't fail when it shouldn't

**Comment by faytey** [comment]
So maybe I got the task wrong, I was asked to handle errors and replace all asserts, also to use the Result method for error handling. Please can you help explain the task properly so I do the correct thing please thanks

**Comment by rcatalan98** [comment]
> So maybe I got the task wrong, I was asked to handle errors and replace all asserts, also to use the Result method for error handling. Please can you help explain the task properly so I do the correct thing please thanks

Hi @faytey! No problem maybe we weren't clear enough with the issue. The task is to refactor the listed function to return `Result` instead of having the assert in this library functions. So for `testing` this functions you will receive whatever the function returns and use patter matching to check if there was an error or not. This same strategy should be used by any function that calls the underlying library functions.

Is that clear?

**Pull request closed without merging**